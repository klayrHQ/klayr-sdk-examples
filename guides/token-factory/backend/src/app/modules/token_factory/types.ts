import { JSONObject } from 'klayr-sdk';

export type TokenID = Buffer;

export interface ModuleConfig {
	maxNameLength: number;
	maxSymbolLength: number;
	maxTotalSupply: bigint;
	minAmountToMint: bigint;
	maxAmountToMint: bigint;
	chainID: Buffer;
	createTokenFee: bigint;
}

export type ModuleConfigJSON = JSONObject<ModuleConfig>;
