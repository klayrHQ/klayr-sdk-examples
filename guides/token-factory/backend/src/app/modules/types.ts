import { JSONObject } from 'lisk-sdk';

export interface ModuleConfig {
	maxNameLength: number;
	maxSymbolLength: number;
	maxTotalSupply: number;
}

export type ModuleConfigJSON = JSONObject<ModuleConfig>;
