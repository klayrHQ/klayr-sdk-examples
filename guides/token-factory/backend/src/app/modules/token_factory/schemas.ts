export const mintSchema = {
	$id: 'token_factory/mintParams',
	title: 'Mints extra tokens to a recipient',
	type: 'object',
	properties: {
		tokenID: {
			dataType: 'bytes',
			fieldNumber: 1,
		},
		amount: {
			dataType: 'uint64',
			fieldNumber: 2,
		},
		recipient: {
			dataType: 'bytes',
			fieldNumber: 3,
			format: 'klayr32',
		},
	},
	required: ['amount', 'recipient'],
};

export const createTokenSchema = {
	$id: 'token_factory/createToken-params',
	title: 'CreateTokenCommand transaction parameter for the Token Factory module',
	type: 'object',
	required: ['name', 'symbol', 'totalSupply'],
	properties: {
		name: {
			dataType: 'string',
			fieldNumber: 1,
			minLength: 1,
			maxLength: 256,
		},
		symbol: {
			dataType: 'string',
			fieldNumber: 2,
			minLength: 1,
			maxLength: 256,
		},
		totalSupply: {
			dataType: 'uint64',
			fieldNumber: 3,
		},
	},
};

export const configSchema = {
	$id: '/token_factory/config',
	type: 'object',
	properties: {
		maxNameLength: {
			dataType: 'uint32',
			fieldNumber: 1,
		},
		maxSymbolLength: {
			dataType: 'uint32',
			fieldNumber: 2,
		},
		maxTotalSupply: {
			dataType: 'uint64',
			fieldNumber: 3,
		},
		minAmountToMint: {
			dataType: 'uint64',
			fieldNumber: 4,
		},
		maxAmountToMint: {
			dataType: 'uint64',
			fieldNumber: 5,
		},
		createTokenFee: {
			dataType: 'uint64',
			fieldNumber: 6,
		},
	},
	required: [
		'maxNameLength',
		'maxSymbolLength',
		'maxTotalSupply',
		'minAmountToMint',
		'maxAmountToMint',
		'createTokenFee',
	],
};
