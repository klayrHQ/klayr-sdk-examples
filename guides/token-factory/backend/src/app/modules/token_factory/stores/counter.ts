import { BaseStore } from 'lisk-sdk';

export interface CounterStoreData {
	counter: bigint;
}

export const counterKey = Buffer.alloc(0);

export const counterStoreSchema = {
	$id: '/token_factory/counter',
	type: 'object',
	required: ['counter'],
	properties: {
		counter: {
			dataType: 'uint64',
			fieldNumber: 1,
		},
	},
};

export class CounterStore extends BaseStore<CounterStoreData> {
	public schema = counterStoreSchema;
}
