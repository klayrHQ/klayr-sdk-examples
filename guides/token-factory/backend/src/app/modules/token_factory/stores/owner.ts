import { BaseStore } from 'klayr-sdk';

export interface OwnerStoreData {
	address: Buffer;
}

export const ownerStoreSchema = {
	$id: '/token_factory/owner',
	type: 'object',
	required: ['address'],
	properties: {
		address: {
			dataType: 'bytes',
			fieldNumber: 1,
			format: 'klayr32',
		},
	},
};

export class OwnerStore extends BaseStore<OwnerStoreData> {
	public schema = ownerStoreSchema;
}
