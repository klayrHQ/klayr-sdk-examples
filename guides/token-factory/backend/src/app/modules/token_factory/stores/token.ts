import { BaseStore } from 'klayr-sdk';

export interface TokenStoreData {
	name: string;
	symbol: string;
	totalSupply: bigint;
}

export const tokenStoreSchema = {
	$id: '/token_factory/token',
	type: 'object',
	required: ['name', 'symbol', 'totalSupply'],
	properties: {
		name: {
			dataType: 'string',
			fieldNumber: 1,
		},
		symbol: {
			dataType: 'string',
			fieldNumber: 2,
		},
		totalSupply: {
			dataType: 'uint64',
			fieldNumber: 3,
		},
	},
};

export class TokenStore extends BaseStore<TokenStoreData> {
	public schema = tokenStoreSchema;
}
