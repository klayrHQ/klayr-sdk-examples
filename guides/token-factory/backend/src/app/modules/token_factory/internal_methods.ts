import { BaseMethod, CommandExecuteContext } from 'klayr-sdk';
import { OwnerStore } from './stores/owner';
import { CounterStore, CounterStoreData, counterKey } from './stores/counter';
import { TokenStore } from './stores/token';
import { CreateTokenParams } from './commands/create_token_command';

export class InternalMethod extends BaseMethod {
	private _chainID!: Buffer;

	public async init(chainID: Buffer): Promise<void> {
		this._chainID = chainID;
	}

	public async getNextTokenIDAndSetStores(
		context: CommandExecuteContext<CreateTokenParams>,
		senderAddress: Buffer,
	) {
		const tokenStore = this.stores.get(TokenStore);
		const counterStore = this.stores.get(CounterStore);
		const ownerStore = this.stores.get(OwnerStore);

		let tokenIdCounter: CounterStoreData = await counterStore
			.get(context, counterKey)
			.catch(() => ({ counter: 0 }));
		tokenIdCounter.counter++;

		const newLocalIDBuffer = Buffer.alloc(4);
		newLocalIDBuffer.writeUInt32BE(tokenIdCounter.counter);
		const newTokenIDBuffer = Buffer.concat([this._chainID, newLocalIDBuffer]);

		await Promise.all([
			counterStore.set(context, counterKey, tokenIdCounter),
			ownerStore.set(context, newTokenIDBuffer, { address: senderAddress }),
			tokenStore.set(context, newTokenIDBuffer, {
				name: context.params.name,
				symbol: context.params.symbol,
				totalSupply: BigInt(context.params.totalSupply),
			}),
		]);

		return newTokenIDBuffer;
	}
}
