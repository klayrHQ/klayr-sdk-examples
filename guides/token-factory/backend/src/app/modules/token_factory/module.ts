/* eslint-disable class-methods-use-this */
/* eslint-disable @typescript-eslint/member-ordering */

import { validator } from '@klayr/validator';
import {
	BaseModule,
	FeeMethod,
	ModuleInitArgs,
	ModuleMetadata,
	TokenMethod,
	utils,
} from 'klayr-sdk';
import { BatchTransferCommand } from './commands/batch_transfer_command';
import { ModuleConfig, ModuleConfigJSON } from './types';
import { BurnCommand } from './commands/burn_command';
import { CreateTokenCommand } from './commands/create_token_command';
import { MintCommand } from './commands/mint_command';
import { TokenFactoryEndpoint } from './endpoint';
import { NewTokenEvent } from './events/new_token';
import { InternalMethod } from './internal_methods';
import { TokenFactoryMethod } from './method';
import { configSchema } from './schemas';
import { CounterStore } from './stores/counter';
import { OwnerStore } from './stores/owner';
import { TokenStore } from './stores/token';
import { getModuleConfig } from './utils';

export const defaultConfig = {
	maxNameLength: 30,
	maxSymbolLength: 5,
	maxTotalSupply: BigInt(1e18), // max: 9223372036854775807 = 9e18
	minAmountToMint: BigInt(1000),
	maxAmountToMint: BigInt(1e6) * BigInt(1e8),
	createTokenFee: BigInt(100_000),
	minAmountToBurn: BigInt(1000),
};

export class TokenFactoryModule extends BaseModule {
	private readonly _internalMethod = new InternalMethod(this.stores, this.events);

	private _createTokenCommand = new CreateTokenCommand(this.stores, this.events);
	private _batchTransferCommand = new BatchTransferCommand(this.stores, this.events);
	private _mintCommand = new MintCommand(this.stores, this.events);
	private _burnCommand = new BurnCommand(this.stores, this.events);

	private _moduleConfig!: ModuleConfig;
	private _tokenMethod!: TokenMethod;
	private _feeMethod!: FeeMethod;

	public endpoint = new TokenFactoryEndpoint(this.stores, this.offchainStores);
	public tokenFactoryMethod = new TokenFactoryMethod(this.stores, this.events);
	public commands = [
		this._createTokenCommand,
		this._mintCommand,
		this._burnCommand,
		this._batchTransferCommand,
	];

	public constructor() {
		super();
		// registeration of stores and events
		this.stores.register(TokenStore, new TokenStore(this.name, 0));
		this.stores.register(CounterStore, new CounterStore(this.name, 1));
		this.stores.register(OwnerStore, new OwnerStore(this.name, 2));

		this.events.register(NewTokenEvent, new NewTokenEvent(this.name));
	}

	public addDependencies(tokenMethod: TokenMethod, feeMethod: FeeMethod) {
		this._tokenMethod = tokenMethod;
		this._feeMethod = feeMethod;
		this._createTokenCommand.addDependencies({
			internalMethod: this._internalMethod,
			tokenMethod: this._tokenMethod,
			feeMethod: this._feeMethod,
		});
		this._mintCommand.addDependencies({
			tokenFactoryMethod: this.tokenFactoryMethod,
			tokenMethod: this._tokenMethod,
		});
		this._burnCommand.addDependencies({
			tokenFactoryMethod: this.tokenFactoryMethod,
			tokenMethod: this._tokenMethod,
		});
		this._batchTransferCommand.addDependencies({
			tokenMethod: this._tokenMethod,
		});
	}

	public metadata(): ModuleMetadata {
		return {
			...this.baseMetadata(),
			commands: this.commands.map(command => ({
				name: command.name,
				params: command.schema,
			})),
			events: this.events.values().map(v => ({
				name: v.name,
				data: v.schema,
			})),
			endpoints: [],
			assets: [],
		};
	}

	// Lifecycle hooks
	public async init(args: ModuleInitArgs): Promise<void> {
		// Get the module config defined in the config.json of the node
		const { moduleConfig, genesisConfig } = args;
		// Overwrite the default module config with values from config.json, if set
		const config = utils.objects.mergeDeep({}, defaultConfig, moduleConfig) as ModuleConfigJSON;
		// Validate the config with the config schema

		validator.validate<ModuleConfigJSON>(configSchema, config);

		this._moduleConfig = getModuleConfig(config, genesisConfig);
		this._internalMethod.init(this._moduleConfig.chainID);
		this._createTokenCommand.init(this._moduleConfig).catch(err => {
			console.log('Error: ', err);
		});
		this._mintCommand
			.init({
				minAmountToMint: this._moduleConfig.minAmountToMint,
				maxAmountToMint: this._moduleConfig.maxAmountToMint,
			})
			.catch(err => {
				console.log('Error: ', err);
			});
		this._burnCommand
			.init({
				minAmountToBurn: this._moduleConfig.minAmountToBurn,
			})
			.catch(err => {
				console.log('Error: ', err);
			});
	}

	// public async insertAssets(_context: InsertAssetContext) {
	// 	// initialize block generation, add asset
	// }

	// public async verifyAssets(_context: BlockVerifyContext): Promise<void> {
	// 	// verify block
	// }

	// Lifecycle hooks
	// public async verifyTransaction(_context: TransactionVerifyContext): Promise<VerificationResult> {
	// verify transaction will be called multiple times in the transaction pool
	// return { status: VerifyStatus.OK };
	// }

	// public async beforeCommandExecute(_context: TransactionExecuteContext): Promise<void> {
	// }

	// public async afterCommandExecute(_context: TransactionExecuteContext): Promise<void> {

	// }
	// public async initGenesisState(_context: GenesisBlockExecuteContext): Promise<void> {

	// }

	// public async finalizeGenesisState(_context: GenesisBlockExecuteContext): Promise<void> {

	// }

	// public async beforeTransactionsExecute(_context: BlockExecuteContext): Promise<void> {

	// }

	// public async afterTransactionsExecute(_context: BlockAfterExecuteContext): Promise<void> {

	// }
}
