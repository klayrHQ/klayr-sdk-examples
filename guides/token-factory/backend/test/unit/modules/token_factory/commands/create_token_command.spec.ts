import {
	CreateTokenCommand,
	CreateTokenParams,
} from '@app/modules/token_factory/commands/create_token_command';
import { TokenFactoryModule } from '@app/modules/token_factory/module';
import { createTokenSchema as createSchema } from '@app/modules/token_factory/schemas';
import { ModuleConfig } from '@app/modules/token_factory/types';
import { CommandExecuteContext, Transaction, VerifyStatus, chain, codec, db } from 'klayr-sdk';
import { TokenID, createCtx, createSampleTransaction } from '@test/helpers';
import { TokenStore } from '@app/modules/token_factory/stores/token';
import { CounterStore, counterKey } from '@app/modules/token_factory/stores/counter';
import { OwnerStore } from '@app/modules/token_factory/stores/owner';
import { InternalMethod } from '@app/modules/token_factory/internal_methods';

describe('CreateTokenCommand', () => {
	const initConfig = {
		maxNameLength: 30,
		maxSymbolLength: 6,
		maxTotalSupply: BigInt(1e6),
		chainID: Buffer.from('12345678'),
		createTokenFee: BigInt(100_000),
	};
	const defaultToken = {
		name: 'The real pepe coin',
		symbol: 'PEPE',
		totalSupply: BigInt(1e4),
	};
	const defaultValidParams = codec.encode(createSchema, defaultToken);
	const mockMint = jest.fn();
	const mockInitialize = jest.fn();
	const mockPayFee = jest.fn();

	let command: CreateTokenCommand;
	let stateStore: any;
	let counterStore: CounterStore;
	let tokenStore: TokenStore;
	let ownerStore: OwnerStore;

	beforeEach(async () => {
		const tokenFactory = new TokenFactoryModule();
		const internalMethod = new InternalMethod(tokenFactory.stores, tokenFactory.events);
		await internalMethod.init(initConfig.chainID);

		command = new CreateTokenCommand(tokenFactory.stores, tokenFactory.events);
		command.addDependencies({
			internalMethod,
			tokenMethod: { mint: mockMint, initializeToken: mockInitialize },
			feeMethod: { payFee: mockPayFee },
		} as any);
		await command.init(initConfig as ModuleConfig);

		stateStore = new chain.StateStore(new db.InMemoryDatabase());
		tokenStore = tokenFactory.stores.get(TokenStore);
		counterStore = tokenFactory.stores.get(CounterStore);
		ownerStore = tokenFactory.stores.get(OwnerStore);
	});

	describe('constructor', () => {
		it('should have valid name', () => {
			expect(command.name).toEqual('createToken');
		});

		it('should have valid schema', () => {
			expect(command.schema).toMatchSnapshot();
		});
	});

	describe('verify', () => {
		describe('schema validation', () => {
			it('should throw when `totalSupply` is too high', async () => {
				const paramWithInvalidTotalSupply = codec.encode(createSchema, {
					name: 'The real pepe coin',
					symbol: 'PEPE',
					totalSupply: initConfig.maxTotalSupply + BigInt(1), // invalid totalSupply
				});
				const transaction = new Transaction(
					createSampleTransaction(paramWithInvalidTotalSupply, CreateTokenCommand.name),
				);
				const ctx = createCtx<CreateTokenParams>(stateStore, transaction, createSchema, 'verify');

				const result = await command.verify(ctx);
				expect(result.status).toBe(VerifyStatus.FAIL);
				expect(result.error).toEqual(new Error('Total supply cannot be greater than 1000000'));
			});

			it('should throw when `fee` is too low', async () => {
				const invalidFee = BigInt(10_000);
				const sender = '3bb9a44b71c83b95045486683fc198fe52dcf27b55291003590fcebff0a45d9a';
				const transaction = new Transaction(
					createSampleTransaction(defaultValidParams, CreateTokenCommand.name, sender, invalidFee),
				);
				const context = createCreateTokenCtx(stateStore, transaction, 'verify');

				const result = await command.verify(context);
				expect(result.status).toBe(VerifyStatus.FAIL);
				expect(result.error).toEqual(new Error('Insufficient transaction fee'));
			});

			it('should be ok when valid params', async () => {
				const transaction = new Transaction(
					createSampleTransaction(defaultValidParams, CreateTokenCommand.name),
				);
				const ctx = createCtx<CreateTokenParams>(stateStore, transaction, createSchema, 'verify');

				const result = await command.verify(ctx);
				expect(result.status).toBe(VerifyStatus.OK);
			});
		});
	});

	describe('execute', () => {
		describe('valid cases', () => {
			it('mint function should have been called', async () => {
				const transaction = new Transaction(
					createSampleTransaction(defaultValidParams, CreateTokenCommand.name),
				);
				const ctx = createCtx<CreateTokenParams>(stateStore, transaction, createSchema, 'execute');

				await expect(
					command.execute(ctx as CommandExecuteContext<CreateTokenParams>),
				).resolves.toBeUndefined();
				expect(mockMint).toHaveBeenCalledTimes(1);
				expect(mockInitialize).toHaveBeenCalledTimes(1);
				expect(mockPayFee).toHaveBeenCalledTimes(1);
			});

			it('should update the `token`, `counter` and `owner` store', async () => {
				const transaction = new Transaction(
					createSampleTransaction(defaultValidParams, CreateTokenCommand.name),
				);
				const tokenID = 1;
				const tokenIDBuf = new TokenID(tokenID).toBuffer();

				const ctx = createCtx<CreateTokenParams>(stateStore, transaction, createSchema, 'execute');

				await expect(
					command.execute(ctx as CommandExecuteContext<CreateTokenParams>),
				).resolves.toBeUndefined();

				const token = await tokenStore.get(ctx, tokenIDBuf);
				const tokenIdCounter = await counterStore.get(ctx, counterKey);
				const owner = await ownerStore.get(ctx, tokenIDBuf);

				expect(token.name).toBe(defaultToken.name);
				expect(token.symbol).toBe(defaultToken.symbol);
				expect(token.totalSupply).toBe(defaultToken.totalSupply);

				expect(tokenIdCounter.counter).toBe(tokenID);

				expect(owner.address.equals(transaction.senderAddress)).toBe(true);
			});
		});

		describe('invalid cases', () => {
			// will add when mint function is implemented
			it.todo('should throw error');
		});
	});
});
