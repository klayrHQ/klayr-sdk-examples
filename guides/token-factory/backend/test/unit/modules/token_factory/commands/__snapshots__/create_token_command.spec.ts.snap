// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateTokenCommand constructor should have valid schema 1`] = `
{
  "$id": "token_factory/createToken-params",
  "properties": {
    "name": {
      "dataType": "string",
      "fieldNumber": 1,
      "maxLength": 30,
      "minLength": 1,
    },
    "symbol": {
      "dataType": "string",
      "fieldNumber": 2,
      "maxLength": 6,
      "minLength": 1,
    },
    "totalSupply": {
      "dataType": "uint64",
      "fieldNumber": 3,
    },
  },
  "required": [
    "name",
    "symbol",
    "totalSupply",
  ],
  "title": "CreateTokenCommand transaction parameter for the Token Factory module",
  "type": "object",
}
`;
